bionic/code_references.py

```diff
@@ -0,0 +1,245 @@
+                if op.argval in context.cells:
+                    set_tos(context.cells[op.argval])
+                else:
+                    # This can happen when we have nested functions. The second
+                    # level or further nested functions won't have free variables
+                    # from any preceding function except for the top level
+                    # function. This is because the code context that gives us
+                    # free variables is created from the function variable, which
+                    # we only have for the top level function. We get only the
+                    # code object for any inner functions. Since we can't get the
+                    # code context for inner functions, we use the code context
+                    # of the top level function.
+                    set_tos(op.argval)
+            elif op.opname == "IMPORT_NAME":
+                # This instruction only appears if the code object imports a
+                # module using the import statement. If a user is importing
+                # modules inside a function, they probably don't want to import
+                # the module until the function execution time.
+                message = oneline(
+                    f"""
+                Bionic found a code reference in file ${code.co_filename} at line
+                ${lineno} that imports ${op.argval} module when hashing
+                ${code.co_name}. Changes inside the module will likely not be
+                detected. To detect changes inside the module, import it globally
+                instead.
+
+                You can also suppress this warning by removing the
+                `suppress_bytecode_warnings` override from the `@version`
+                decorator on the corresponding function.
+                """
+                )
+                warnings.warn(message)
+                set_tos(None)
+            elif op.opname in ["LOAD_METHOD", "LOAD_ATTR"]:
+                if isinstance(tos, str):
+                    tos += "." + op.argval
+                # TODO: Consider calling getattr only when TOS is a module.
+                # Doing so has risk of missing cases, and any missing case would
+                # be hard to detect for users. Before making this change,
+                # document all the cases that we won't be catching so that we are
+                # aware of them.
+                # TODO: Due to the same reason as why we don't import modules, we
+                # should not call `getattr` on properties. We should either add a
+                # separate check for properties, or implement the above TODO to
+                # fix this.
+                elif hasattr(tos, op.argval):
+                    tos = getattr(tos, op.argval)
+                else:
+                    refs.append(op.argval)
+            elif op.opname == "STORE_FAST" and tos:
+                context.varnames[op.argval] = tos
+                tos = None
```

Comment:
I think this is because it stores the `tos` in `context.varnames`. But I agree that maybe we should use `set_tos(None)` and return the current `tos` as a reference. This is exactly what I meant to do from this [TODO](https://github.com/square/bionic/pull/232/files#diff-9e61581018162d3520e8e837df29a47bb3bea1afc6e11643b25b66d4e67d4a03R254).